.github/workflows/node.yml

```diff
@@ -38,37 +55,86 @@ env:
     CARGO_TERM_COLOR: always
 
 jobs:
-    load-engine-matrix:
+    check_should_run:
       runs-on: ubuntu-latest
       outputs:
-        matrix: ${{ steps.load-engine-matrix.outputs.matrix }}
+        should_run: ${{ steps.check_should_run.outputs.should_run }}
       steps:
-          - name: Checkout
-            uses: actions/checkout@v4
+        - name: Check for Core changes label
+          id: check_should_run
+          run: |
+            if [[ ("${{ github.event.label.name }}" == "Core changes" &&  "${{github.event_name}}" == 'pull_request_target') || "${{github.event_name}}" != 'pull_request_target' ]]; then
+              echo "should_run=true" >> $GITHUB_OUTPUT
+            else
+              echo "should_run=false" >> $GITHUB_OUTPUT
+            fi
+    load-engine-matrix:
+        runs-on: ubuntu-latest
+        # We want to run this job only if the previous job outputs should_run=true,
+        # Which means that if the trigger was a label we run this job just if the label was Core changes
+        needs: check_should_run
+        if: needs.check_should_run.outputs.should_run == 'true'
+        outputs:
+          matrix: ${{ steps.load-engine-matrix.outputs.matrix }}
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4
+
+            - name: Load the engine matrix
+              id: load-engine-matrix
+              shell: bash
+              # We want to run the full matrix only for scheduled runs and Core changes
+              # For other events we want to run only the always matrix which is first and last version supported
+              run: |
+                if [[ "${{ github.event_name }}" == "pull_request" || "${{ github.event_name }}" == "push" ]]; then
+                  echo "matrix=$(jq -c '[.[] | select(.run == "always")]' < .github/json_matrices/engine-matrix.json)" >> $GITHUB_OUTPUT
+                else
+                  echo "matrix=$(jq -c . < .github/json_matrices/engine-matrix.json)" >> $GITHUB_OUTPUT
+                fi
+    create-version-matrix:
+        runs-on: ubuntu-latest
+        needs: check_should_run
+        if: needs.check_should_run.outputs.should_run == 'true'
+        outputs:
+            matrix: ${{ steps.create-version-matrix.outputs.matrix }}
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4
+
+            - name: Create version matrix
+              id: create-version-matrix
+              shell: bash
+              # We want to run the full matrix only for scheduled runs and Core changes
+              # For other events we want to run only the always matrix which is first and last version supported
+              run: |
+                if [[ "${{ github.event_name }}" == "pull_request" || "${{ github.event_name }}" == "push" ]]; then
+                  echo 'matrix={"node":["16.x","20.x"]}' >> $GITHUB_OUTPUT
```

Comment:
Maybe we need one more matrix with language/framework versions? Consider including `run: "always"` where it is needed.

Reply:
Hamm I don't think so, the logic is not the same for the different languages, and generally speaking if it's not helping avoid repetitive code or add simplicity in some way, putting all logic in one place and making the logic straight forward seems better to me
understanding JSON parsing is harder than creating a matrix and avoiding moving between files in order to understand code is better